File: sdks/python/src/opik/evaluation/metrics/heuristics/is_json.py
import json
from typing import Any

from .. import base_metric, score_result


class IsJson(base_metric.BaseMetric):
    """
    A metric that checks if a given output string is valid JSON.

    This metric returns a score of 1.0 if the output string can be parsed as JSON,
    and 0.0 otherwise.

    Args:
        name: The name of the metric. Defaults to "is_json_metric".

    Example:
        >>> from comet_llm_eval.evaluation.metrics import IsJson
        >>> is_json_metric = IsJson()
        >>> result = is_json_metric.score('{"key": "value"}')
        >>> print(result.value)
        1.0
        >>> result = is_json_metric.score('Not a JSON string')
        >>> print(result.value)
        0.0
    """

    def __init__(self, name: str = "is_json_metric") -> None:
        super().__init__(name)

    def score(self, output: str, **ignored_kwargs: Any) -> score_result.ScoreResult:
        """
        Calculate the score based on whether the output string is valid JSON.

        Args:
            output: The output string to check for JSON validity.
            **ignored_kwargs: Additional keyword arguments that are ignored.

        Returns:
            score_result.ScoreResult: A ScoreResult object with a value of 1.0 if the output
                is valid JSON, 0.0 otherwise.
        """
        try:
            json.loads(output)
            return score_result.ScoreResult(value=1.0, name=self.name)
        except Exception:
            return score_result.ScoreResult(value=0.0, name=self.name)

File: sdks/typescript/src/opik/evaluation/metrics/heuristics/IsJson.ts
import { z } from "zod";
import { EvaluationScoreResult } from "../../types";
import { BaseMetric } from "../BaseMetric";

const validationSchema = z.object({
  output: z.unknown(),
});
type Input = z.infer<typeof validationSchema>;

/**
 * IsJson metric - checks if a given output string is valid JSON.
 */
export class IsJson extends BaseMetric {
  /**
   * Creates a new IsJson metric
   * @param name Optional name for the metric (defaults to "is_json_metric")
   * @param trackMetric Whether to track the metric
   */
  constructor(name = "is_json_metric", trackMetric = true) {
    super(name, trackMetric);
  }

  public validationSchema = validationSchema;

  /**
   * Calculates a score based on whether output is a valid JSON
   * @param input Actual output to evaluate
   * @returns Score result (1.0 if valid json, 0.0 if not valid json)
   */
  async score(input: Input): Promise<EvaluationScoreResult> {
    const { output } = input;

    try {
      JSON.parse(output as string);
      return {
        name: this.name,
        value: 1.0,
        reason: "IsJson: Output is valid JSON.",
      };
    } catch {
      return {
        name: this.name,
        value: 0.0,
        reason: "IsJson: Output is not valid JSON",
      };
    }
  }
}
